lib/screens/screens_by_alert/memcache.ex

```diff
@@ -1,29 +1,176 @@
 defmodule Screens.ScreensByAlert.Memcache do
   @moduledoc """
-  ScreensByAlert backend which uses Memcache as a backend.
+  ScreensByAlert backend which uses memcached as a backend.
+
+  This is intended to be used as a singleton process--all public functions in the module
+  communicate with a named process that maintains a connection with memcached.
+
+  Structure of cached data:
+  ```
+  # A Unix timestamp
+  @type timestamp :: integer()
+
+  @type screen_id :: String.t()
+  @type alert_id :: String.t()
+
+  @type timestamped_screen_id :: {screen_id, timestamp}
+
+  %{
+    # Each alert is mapped to a list of screens showing it.
+    # Each screen in the list is paired with a timestamp of the last
+    # time it declared that it was showing this alert.
+    "screens_by_alert." <> alert_id => list(timestamped_screen_id),
+
+    # Metadata to make the "self-refresh" mechanism possible
+    "screens_last_updated." <> screen_id => timestamp
+  }
+  ```
   """
+  alias Screens.ScreensByAlert.Memcache.TaskSupervisor
+
+  require Logger
 
   @behaviour Screens.ScreensByAlert.Behaviour
 
+  @server __MODULE__.Server
+
+  @screens_by_alert_key_prefix "screens_by_alert."
+  @screens_last_updated_key_prefix "screens_last_updated."
+
+  @config Application.compile_env!(:screens, :screens_by_alert)
+  @screens_by_alert_ttl Keyword.fetch!(@config, :screens_by_alert_ttl_seconds)
+  @screens_last_updated_ttl Keyword.fetch!(@config, :screens_last_updated_ttl_seconds)
+  @screens_ttl Keyword.fetch!(@config, :screens_ttl_seconds)
+
   @impl true
-  def start_link(_opts \\ []) do
-    {:ok, config} = Application.fetch_env(:screens, ScreensByAlert.Memcache)
+  def start_link(_opts) do
+    module_config = Application.fetch_env!(:screens, Screens.ScreensByAlert.Memcache)
+    connection_opts = Keyword.fetch!(module_config, :connection_opts)
 
-    Memcache.start_link(config[:connection_opts])
+    Memcache.start_link(connection_opts, name: @server)
   end
 
   @impl true
-  def put_data(_screen_id, _alert_ids) do
+  def put_data(screen_id, alert_ids) when is_binary(screen_id) and is_list(alert_ids) do
```

Comment:
Does there need to be a definition without the guard to avoid uncaught errors, or is that handled?

Reply:
The guards can be thought of sort of like type annotations on the arguments.

Roughly equivalent TypeScript code:
```ts
function putData(screenID: string, alertIDs: string[]): void {...}
```

In TS, we wouldn't expect to need to handle other input types in `putData`—if another function called `putData("screen1", "alert1")`, your editor would underline it and point out the mistake.